numpy/_core/src/multiarray/stringdtype/casts.cpp

```diff
@@ -19,36 +23,243 @@
 #include "dtype.h"
 #include "utf8_utils.h"
 
+#include "casts.h"
+
+
+// Get the shortname letter associated with the given type number.
+NPY_TYPECHAR typenum_to_typechar(NPY_TYPES typenum) {
+    switch (typenum) {
+        case NPY_BOOL:
+            return NPY_BOOLLTR;
+        case NPY_BYTE:
+            return NPY_BYTELTR;
+        case NPY_UBYTE:
+            return NPY_UBYTELTR;
+        case NPY_SHORT:
+            return NPY_SHORTLTR;
+        case NPY_USHORT:
+            return NPY_USHORTLTR;
+        case NPY_INT:
+            return NPY_INTLTR;
+        case NPY_UINT:
+            return NPY_UINTLTR;
+        case NPY_LONG:
+            return NPY_LONGLTR;
+        case NPY_ULONG:
+            return NPY_ULONGLTR;
+        case NPY_LONGLONG:
+            return NPY_LONGLONGLTR;
+        case NPY_ULONGLONG:
+            return NPY_ULONGLONGLTR;
+        case NPY_HALF:
+            return NPY_HALFLTR;
+        case NPY_FLOAT:
+            return NPY_FLOATLTR;
+        case NPY_DOUBLE:
+            return NPY_DOUBLELTR;
+        case NPY_LONGDOUBLE:
+            return NPY_LONGDOUBLELTR;
+        case NPY_CFLOAT:
+            return NPY_CFLOATLTR;
+        case NPY_CDOUBLE:
+            return NPY_CDOUBLELTR;
+        case NPY_CLONGDOUBLE:
+            return NPY_CLONGDOUBLELTR;
+        case NPY_OBJECT:
+            return NPY_OBJECTLTR;
+        case NPY_STRING:
+            return NPY_STRINGLTR;
+        case NPY_UNICODE:
+            return NPY_UNICODELTR;
+        case NPY_VOID:
+            return NPY_VOIDLTR;
+        case NPY_DATETIME:
+            return NPY_DATETIMELTR;
+        case NPY_TIMEDELTA:
+            return NPY_TIMEDELTALTR;
+        case NPY_CHAR:
+            return NPY_CHARLTR;
+        case NPY_VSTRING:
+            return NPY_VSTRINGLTR;
+        default:
+            PyErr_SetString(
+                PyExc_TypeError,
+                "No NPY_TYPECHAR associated with the given typenum."
+            );
+            return static_cast<NPY_TYPECHAR>('\0');
+    }
+}
+
+// Get a c string representation of a type number.
+const char *typenum_to_cstr(NPY_TYPES typenum) {
```

Comment:
I feel like this should be in a NumPy utility header somewhere rather than in this file but don't have a strong opinion for where it makes more sense.

Also in NumPy generally we use `static` functions for any functions that are only needed in a single compilation unit. If it needs to be added to a header, then it needs to be annotated with `NPY_NO_EXPORT`.

This function should be `static const char *`.

Also generally in NumPy we put the type of a function and then a newline and then the function signature.

```suggestion
static const char *
typenum_to_cstr(NPY_TYPES typenum) {
```